Parse Benchmark Prices - Aptos Contract Implementation

hhdgknsn opened the pull request:
**Description:**

This pull request introduces new functionality to the Pyth Aptos contract to parse and validate benchmark prices, similar to the parsePriceFeedUpdates method on the EVM chain. The new methods allow the contract to verify and consume benchmark price data from off-chain API endpoints, ensuring they meet specific publish time boundaries and include all requested price IDs.

**Added Functions Overview:**

`parse_and_validate_price_feeds`:
- Validates publish times and populates a vector of parsed price feeds.
- Ensures all requested price IDs are included in the parsed results.

`parse_price_feed_updates_single_vaa`:
- Parses a single VAA and returns a vector of valid price feeds.

`parse_price_feed_updates`:
- Parses multiple VAAs, validates, and aggregates benchmark prices.
- Handles fee validation and ensures all price updates are within the specified time range.

This new pull request replaces the initial one I have closed, reflecting significant changes in the functionality and logic of the benchmark price parsing methods. The updated implementation now includes enhanced logic for processing price updates, integrated fee validation and handling, and the addition of working tests to ensure reliability and correctness of the new features. These improvements address issues from the previous implementation and provide a more robust solution for the Pyth Aptos contract.

Additionally, the assertion statement - `assert!(is_coin_initialized<CoinType>(), error::invalid_argument(ECOIN_INFO_NOT_PUBLISHED));` (inside the `is_account_registered` function in the coin.move file (aptos-move/framework/aptos-framework/sources)) was causing some of the original contract method tests to fail, including my own added tests. After commenting out this line temporarily, all tests are functioning correctly.

Please review the latest commit and let me know if any further amendments are needed.

Thanks!

m30m commented on target_chains/aptos/contracts/sources/pyth.move:
```diff
@@ -395,6 +394,119 @@ module pyth::pyth {
+
+    public fun parse_price_feed_updates(
+        update_data: vector<vector<u8>>,
+        price_ids: vector<vector<u8>>,
```
In the new version of pythnet, the user can construct the message with only the price_ids they require. So there is no need to accept this argument and we can remove the checks of price_ids all together

m30m commented on target_chains/aptos/contracts/sources/pyth.move:
```diff
@@ -395,6 +394,119 @@ module pyth::pyth {
+        while(i < vector::length(price_infos)) {
+            let price_info = vector::borrow(price_infos, i);
+            let price_feed = price_info::get_price_feed(price_info);
+            let price: price::Price = price_feed::get_price(price_feed);
```
no need to define the type

m30m commented on target_chains/aptos/contracts/sources/pyth.move:
```diff
@@ -395,6 +394,119 @@ module pyth::pyth {
+        fee: Coin<AptosCoin>
+    ): vector<price_feed::PriceFeed> {
+        // validate and deposit the update fee 
+        let update_fee = get_update_fee(&update_data);
```
this will run all the verification again. You can assert the fees were enough at the end of the function using the length of the price_feeds. Same as what we have done on `update_price_feeds`

m30m commented on target_chains/aptos/contracts/sources/pyth.move:
```diff
@@ -1430,4 +1542,179 @@ module pyth::pyth_test {
+
+    #[test(aptos_framework = @aptos_framework)]
+    #[expected_failure(abort_code = 393224, location = pyth::pyth)]
+    fun test_parse_price_feed_updates_invalid_publish_times(aptos_framework: &signer) {
```
we should have a separate test where publish times are valid but outside the price times

m30m commented on target_chains/aptos/contracts/sources/pyth.move:
```diff
@@ -1430,4 +1542,179 @@ module pyth::pyth_test {
+
+    
+    #[test(aptos_framework = @aptos_framework)]
+    fun test_parse_price_feed_updates_success(aptos_framework: &signer) {
```
add assertions for the transfer of fees

hhdgknsn commented:
**Summary of Changes**
I have made the following updates to the `parse_benchmark_prices_v3` branch based on the feedback provided:

**1. Removed price_ids Parameter:**  The `price_ids` parameter and associated checks have been removed from the `parse_price_feed_updates` and related functions.
**2. Adjusted Fee Calculation:** The fee calculation in the `parse_price_feed_updates` function has been updated to calculate and charge the message update fee based on the number of price feeds updated, instead of running the verification multiple times. Additionally, the `parse_price_feed_updates_single_vaa` function was modified to return the number of updates along with the parsed price feeds, ensuring accurate fee validation at the end of the function.
**3. Added Fee Transfer Validation:** The test case `test_parse_price_feed_updates_success` has been updated to include assertions that validate the fee transfer to the Pyth contract. This ensures that the fee is correctly deposited.
**4. Validated Publish Times for Price Feeds:** The `parse_and_validate_price_feeds` function has been updated to assert that each price feed's timestamp falls within the specified publish times. If any price feed is outside this range, an `error::price_feed_outside_time_range` is thrown.
**5. Removed Type Definition:** Removed the type definition from the statement `let price = price_feed::get_price(price_feed); ` in the `parse_and_validate_price_feeds` function as per the review comment.
**6. New Test Case for Publish Times Outside Timestamps:** A new test case `test_parse_price_feed_updates_publish_times_outside_timestamps` has been added to validate that no price feeds are parsed if the publish times are outside the range of all price feeds.
**7. Custom Errors for Invalid Publish Time Range:** Two new custom errors have been defined in the pyth::error module. The `invalid_publish_time_range` error is used in `parse_price_feed_updates` to validate that `min_publish_time` is less than or equal to `max_publish_time`. The `price_feed_outside_time_range` error ensures that each price feed's timestamp falls within the specified publish times.

I have also ensured that all relevant test cases reflect these changes to verify the correctness and reliability of the new functionality.

Happy to adjust anything further as required. Many thanks.

m30m commented on target_chains/aptos/contracts/sources/pyth.move:
```diff
@@ -395,6 +394,104 @@ module pyth::pyth {
+        max_publish_time: u64
+    ): vector<price_feed::PriceFeed> {
+        let parsed_price_feeds = vector::empty<price_feed::PriceFeed>();
+        let parsed_price_ids = vector::empty<vector<u8>>();
```
this vector is useless

m30m commented on target_chains/aptos/contracts/sources/pyth.move:
```diff
@@ -1430,4 +1527,158 @@ module pyth::pyth_test {
+        let max_publish_time: u64 = 1663680750;
+
+        let test_price_feeds: vector<pyth::price_feed::PriceFeed> = pyth::parse_price_feed_updates(update_data, min_publish_time, max_publish_time, coins);        
+        assert!(vector::length(&test_price_feeds) == 0, 1);
```
I don't think the test reaches this line since the previous line will fail. We can probably remove it for clarity. the same goes for the next few testcases where we expect failure

hhdgknsn commented:
A summary of the changes I've made:

- Removed the redundant `parsed_price_ids` vector from the `parse_and_validate_price_feeds` function as it was not used or returned.

- Removed redundant assertions that were not reachable due to expected failures in the unit tests.

- Confirmed that all tests pass as expected.
